File: 2024-12-December-GFG-POTD/December 11 - Merge Without Extra Space/Solution.py
class Solution:
    # Function to merge the arrays.
    def mergeArrays(self, a, b):
        n, m = len(a), len(b)
        gap = self.nextGap(n + m)

        while gap > 0:
            i, j = 0, 0

            # Compare and swap within the first array.
            while i + gap < n:
                if a[i] > a[i + gap]:
                    a[i], a[i + gap] = a[i + gap], a[i]
                i += 1

            # Compare and swap between arrays.
            j = gap - n if gap > n else 0
            while i < n and j < m:
                if a[i] > b[j]:
                    a[i], b[j] = b[j], a[i]
                i += 1
                j += 1

            # Compare and swap within the second array.
            while j + gap < m:
                if b[j] > b[j + gap]:
                    b[j], b[j + gap] = b[j + gap], b[j]
                j += 1

            # Calculate the next gap.
            gap = self.nextGap(gap)

    # Function to calculate the next gap.
    def nextGap(self, gap):
        if gap <= 1:
            return 0
        return (gap // 2) + (gap % 2)

# Driver Code
if __name__ == "__main__":
    t = int(input())  # Number of test cases
    for _ in range(t):
        a = list(map(int, input().strip().split()))
        b = list(map(int, input().strip().split()))

        ob = Solution()
        ob.mergeArrays(a, b)

        print(" ".join(map(str, a)))
        print(" ".join(map(str, b)))

File: 2024-12-December-GFG-POTD/December 11 - Merge Without Extra Space/Solution.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

class Solution {
public:
    // Function to merge the arrays.
    void mergeArrays(vector<int>& a, vector<int>& b) {
        int n = a.size(), m = b.size();
        int gap = nextGap(n + m);

        while (gap > 0) {
            int i, j;

            // Compare and swap within the first array.
            for (i = 0; i + gap < n; i++) {
                if (a[i] > a[i + gap]) {
                    swap(a[i], a[i + gap]);
                }
            }

            // Compare and swap between arrays.
            for (j = gap > n ? gap - n : 0; i < n && j < m; i++, j++) {
                if (a[i] > b[j]) {
                    swap(a[i], b[j]);
                }
            }

            // Compare and swap within the second array.
            if (j < m) {
                for (j = 0; j + gap < m; j++) {
                    if (b[j] > b[j + gap]) {
                        swap(b[j], b[j + gap]);
                    }
                }
            }

            // Calculate the next gap.
            gap = nextGap(gap);
        }
    }

private:
    // Function to calculate the next gap.
    int nextGap(int gap) {
        if (gap <= 1) return 0;
        return (gap / 2) + (gap % 2);
    }
};

// Driver Code
int main() {
    int t;
    cin >> t; // Number of test cases
    while (t--) {
        int n, m;
        cin >> n >> m;

        vector<int> a(n), b(m);
        for (int i = 0; i < n; i++) cin >> a[i];
        for (int i = 0; i < m; i++) cin >> b[i];

        Solution ob;
        ob.mergeArrays(a, b);

        for (int i = 0; i < n; i++) cout << a[i] << " ";
        cout << endl;
        for (int i = 0; i < m; i++) cout << b[i] << " ";
        cout << endl;
    }
    return 0;
}
